OptimoveSDK/Sources/Classes/Optimobile/Optimobile+Push.swift

```diff
@@ -404,55 +404,61 @@ class PushHelper {
         let didReceiveSelector = #selector(UIApplicationDelegate.application(_:didReceiveRemoteNotification:fetchCompletionHandler:))
         let receiveType = NSString(string: "v@:@@@?").utf8String
         let didReceive : didReceiveBlock = { (obj:Any, _ application: UIApplication, userInfo: [AnyHashable : Any], completionHandler: @escaping (UIBackgroundFetchResult) -> Void) in
-            var fetchResult : UIBackgroundFetchResult = .noData
-            let fetchBarrier = DispatchSemaphore(value: 0)
-
-            if let _ = existingDidReceive {
-                unsafeBitCast(existingDidReceive, to: kumulos_applicationDidReceiveRemoteNotificationFetchCompletionHandler.self)(obj, didReceiveSelector, application, userInfo, { (result : UIBackgroundFetchResult) in
-                    fetchResult = result
-                    fetchBarrier.signal()
-                })
-            } else {
-                fetchBarrier.signal()
-            }
-            
-            if UIApplication.shared.applicationState == .inactive {
-               if #available(iOS 10, *) {
-                   // Noop (tap handler in delegate will deal with opening the URL)
-               } else {
-                   Optimobile.sharedInstance.pushHandleOpen(withUserInfo:userInfo)
-               }
+            let notification = PushNotification(userInfo: userInfo)
+            let hasInApp = notification.inAppDeepLink() != nil
+
+            self.setBadge(userInfo: userInfo)
+            self.trackPushDelivery(notification: notification)
+
+            if existingDidReceive == nil && !hasInApp {
+                // Nothing to do
+                completionHandler(.noData)
+                return
+            } else if existingDidReceive != nil && !hasInApp {
+                // Only existing delegate work to do
+                unsafeBitCast(existingDidReceive, to: kumulos_applicationDidReceiveRemoteNotificationFetchCompletionHandler.self)(obj, didReceiveSelector, application, userInfo, completionHandler)
+                return
             }
-            
-            let aps = userInfo["aps"] as! [AnyHashable:Any]
-            guard let contentAvailable = aps["content-available"] as? Int, contentAvailable == 1 else {
-                if #available(iOS 10, *) {} else {
-                    self.setBadge(userInfo: userInfo)
-                    self.trackPushDelivery(userInfo: userInfo)
+
+            var fetchResult : UIBackgroundFetchResult = .noData
+            let group = DispatchGroup()
+
+            if existingDidReceive != nil {
+                group.enter()
+                DispatchQueue.main.async {
+                    unsafeBitCast(existingDidReceive, to: kumulos_applicationDidReceiveRemoteNotificationFetchCompletionHandler.self)(obj, didReceiveSelector, application, userInfo, { (result : UIBackgroundFetchResult) in
+                        DispatchQueue.main.async {
+                            if fetchResult == .noData {
```

Comment:
Result handling is also sent back to the main thread to ensure thread-safety mutating the var